src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4574,22 +4576,42 @@ private NullableFlowState LearnFromCompareExchangeMethod(in CompareExchangeInfo
             //     location = value;
             // }
 
-            var comparand = compareExchangeInfo.Arguments[2];
-            var valueFlowState = compareExchangeInfo.Results[1].RValueType.State;
+            Debug.Assert(compareExchangeInfo.Arguments.Length == 3);
+
+            var comparand = compareExchangeInfo.Arguments[GetArgumentOrdinalFromParameterOrdinal(2, compareExchangeInfo.ArgsToParamsOpt)];
```

Comment:
Consider using an approach like the following

```cs
var argsToParamsOpt = compareExchangeInfo.ArgsToParamsOpt;
Debug.Assert(argsToParamsOpt is { IsDefault: true } or { Length: 3 });
var (comparandIndex, valueIndex, locationIndex) = argsToParamsOpt.IsDefault
    ? (2, 1, 0)
    : (argsToParamsOpt.IndexOf(2), argsToParamsOpt.IndexOf(1), argsToParamsOpt.IndexOf(0));
```

then update indexer arguments to use comparandIndex instead of 2, etc.